backend/gadget.pm

```diff
@@ -0,0 +1,139 @@
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+# this backend uses a KVM connector speaking VNC and external tools
+# for serial line and power cycling
+
+package backend::gadget;
+use strict;
+use base ('backend::baseclass');
+use bmwqemu qw(diag);
+use testapi qw(get_required_var get_var);
+use IPC::Run ();
+require IPC::System::Simple;
+use autodie qw(:all);
+
+sub new {
+    my $class = shift;
+    get_required_var('WORKER_HOSTNAME');
+    return $class->SUPER::new;
+}
+
+sub run_cmd {
+    my ($self, $cmd) = @_;
+
+    my @cmd = split(/ /, $cmd);
+
+    my ($stdin, $stdout, $stderr, $ret);
+    $ret = IPC::Run::run(\@cmd, \$stdin, \$stdout, \$stderr);
+    chomp $stdout;
+    chomp $stderr;
+
+    die join(' ', @cmd) . ": $stderr" unless ($ret);
+    bmwqemu::diag("IPMI: $stdout");
+    return $stdout;
+}
+
+sub poweroff_host {
+    my ($self) = @_;
+    $self->run_cmd(get_required_var('GADGET_POWEROFF_CMD'));
+}
+
+sub restart_host {
+    my ($self) = @_;
+
+    $self->poweroff_host;
+    sleep(3);
+    $self->run_cmd(get_var('GADGET_POWERON_CMD'));
+    return;
+}
+
+sub relogin_vnc {
+    my ($self) = @_;
+
+    if ($self->{vnc}) {
+        close($self->{vnc}->socket);
+        sleep(1);
+    }
+
+    my $vnc = $testapi::distri->add_console(
+        'sut',
+        'vnc-base',
+        {
+            hostname        => get_required_var('GADGET_VNC_IP'),
+            port            => 5900,
+            depth           => 16,
+            connect_timeout => 50
+        });
+    $vnc->backend($self);
+    $self->select_console({testapi_console => 'sut'});
+
+    return 1;
+}
+
+sub do_start_vm() {
+    my ($self) = @_;
+
+    # remove backend.crashed
+    $self->unlink_crash_file;
+    $self->restart_host;
+    $self->relogin_vnc;
+    $self->start_serial_grab;
+    return {};
+}
+
+sub do_stop_vm {
+    my ($self) = @_;
+
+    $self->poweroff_host;
+    $self->stop_serial_grab();
+    return {};
+}
+
+sub status {
+    my ($self) = @_;
+    print "status ignored\n";
+    return;
+}
+
+# serial grab
+
+sub start_serial_grab {
+    my ($self) = @_;
+
+    $self->{serialpid} = fork();
+    if ($self->{serialpid} == 0) {
+        setpgrp 0, 0;
+        open(my $serial, '>',  $self->{serialfile});
+        open(STDOUT,     ">&", $serial);
+        open(STDERR,     ">&", $serial);
+        exec(get_required_var('GADGET_SOL_CMD'));
```

Comment:
I have to leave that to @sysrich , as I don't know what possible commands we possibly have to take care of.